DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+# First class functions
+
+| Section        | Value                                  |
+|----------------|----------------------------------------|
+| DIP:           | 27                                     |
+| Author:        | Amaury SÉCHET                          |
+| Status:        |                                        |
+
+## Abstract
+
+This DIP is the first of 3. It intends to define what is a function in D. The DIP propose a radically simple approach by deciding that all function are first class function. The DIP proposes ways to mitigate breakage involved and allow optional parenthesis when no confusion is possible.
+
+## Description
+
+Function are still defined with the same syntax :
+
+```d
+ReturnType functionName(Parameters) {
+  // Function body.
+}
+```
+
+However, this is now strictly equivalent to :
+
+```d
+enum functionName = ReturnType function(Parameters) {
+  // Function body.
+};
+```
+
+The function has a function type and can be used as this anywhere a function is expected :
+
+```d
+static assert(is(typeof(functionName) == ReturnType function(Parameters)); // OK
+
+ReturnType function(Parameters) foo = functionName; // OK
+
+void buzz(ReturnType function(Parameters) qux);
+buzz(foo); // OK
+buzz(bar); // OK
+
+auto a = functionName;
+static assert(is(typeof(a) == ReturnType function(Parameters)); // OK
+
+auto b = functionName();
+static assert(is(typeof(b) == ReturnType); // OK
```

Comment:
Ha I get it now, I did not put any argument in the call. It should be auto b = functionName(Arguments);